cpmpy/transformations/int2bool.py

```diff
@@ -0,0 +1,469 @@
+        return encoding
+    elif _dom_size(x) >= 100:
+        # even slightly large domain sizes suit the binary encoding, as the PB encoding is not expected to be optimized for integer variables!
+        return "binary"
+    elif cmp in (None, "==", "!="):
+        return "direct"  # equalities suit the direct encoding
+    else:  # inequalities suit the order encoding
+        return "order"
+
+
+class IntVarEnc(ABC):
+    """Abstract base class for integer variable encodings."""
+
+    def __init__(self, x, xs):
+        """Create encoding of integer variable `x`."""
+        if _dom_size(x) == 0:
+            raise EMPTY_DOMAIN_ERROR
+        self._x = x  # the encoded integer variable
+        self._xs = xs  # its encoding variables
+
+    def vars(self):
+        """Return the Boolean variables in the encoding."""
+        return self._xs
+
+    def _offset(self, d):
+        """ "Map domain value `d` to the encoding range (e.g. for the direct encoding to the encoding variable index x==i)"""
+        return d - self._x.lb
+
+    def decode(self):
+        """Return the integer value of the encoding."""
+        terms, k = self.encode_term()
+        for weight, lit in terms:
+            if lit is None:
+                return None
+            elif lit.value() is True:
+                k += weight
+        return k
+
+    @abstractmethod
+    def encode_domain_constraint(self):
+        """
+        Return domain constraints for the encoding.
+
+        Returns:
+            List[Expression]: a list of constraints
+        """
+        pass
+
+    @abstractmethod
+    def encode_comparison(self, op, rhs):
+        """
+        Encode a comparison over the variable: self <op> rhs.
+
+        Args:
+            op: The comparison operator ("==", "!=", "<", "<=", ">", ">=")
+            rhs: The right-hand side value to compare against
+
+        Returns:
+            List[Expression]: a list of constraints
+        """
+        pass
+
+    @abstractmethod
+    def encode_term(self, w=1):
+        """Encode `w*self` as a weighted sum of Boolean variables plus a constant.
+
+        Args:
+            w: The weight to multiply the variable by
+
+        Returns:
+            tuple: (terms, k)  # with `terms` a list of `(weight, literal)` tuples and `k` an integer constant
+        """
+        pass
+
+
+class IntVarEncDirect(IntVarEnc):
+    """
+    Direct (or sparse or one-hot) encoding of an integer variable.
+
+    Uses a Boolean 'equality' variable for each value in the domain.
+    """
+
+    def __init__(self, x):
+        """Create direct encoding of integer variable `x`."""
+        name = f"EncDir({x.name}"
+        dom_size = _dom_size(x)
+        if dom_size == 0:
+            raise EMPTY_DOMAIN_ERROR
+        elif dom_size == 1:
+            # TODO we would like to use a boolval(true) here rather than a literal, but this causes problems in linearize
+            xs = cp.cpm_array([boolvar(name=name)])
```

Comment:
why wrap it in a cpm_array?
if `boolvar()` itself does not work, try `boolvar(shape=(1,))`

Reply:
Ahh I was looking for this (because `shape=1` does not return a listlike, to my frustration!). That's very good to know, that will clean up a lot of parts.